project-docs/javascript-linter-spec.md

```diff
@@ -0,0 +1,163 @@
+
+So at a high level, compliance to a recipe means:
+
+_the items in the page must correspond to the ingredients listed in the recipe, in the correct order, and each item must satisfy the particular requirements for its ingredient_.
+
+## JavaScript recipes
+
+We've defined the following recipes for the JS docs:
+
+- [javascript-class](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-class.yaml)
+- [javascript-namespace](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-namespace.yaml)
+- [javascript-constructor](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-constructor.yaml)
+- [javascript-method](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-method.yaml)
+- [javascript-property](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-property.yaml)
+- [javascript-error](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-error.yaml)
+- [javascript-language-feature](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-language-feature.yaml)
+
+## JavaScript recipe ingredients
+
+In total these recipes use the following ingredients:
+
+prose ingredients:
+
+- prose.short_description
+- prose.description?
+- prose.see_also
+- prose.syntax
+- prose.message
+- prose.error_type
+- prose.what_went_wrong
+- prose.\*
+
+data ingredients:
+
+- data.interactive_example?
+- data.class_constructor
+- data.constructor_properties?
+- data.static_properties?
+- data.static_methods?
+- data.instance_properties?
+- data.instance_methods?
+- data.examples
+- data.specifications
+- data.browser_compatibility
+
+All ingredients except two are demarcated in the page using an `<h2>` heading. The exceptions are `prose.short_description` and `data.interactive_example`.
+
+### Prose ingredients
+
+With three exceptions, all prose ingredients impose the same requirement on the page: it must have an `<h2>` section whose ID attribute maps to the named section. The mapping is simply that `prose.` is removed, and the first letter is capitalized:
+
+```
+prose.syntax      ->    Syntax
+prose.see_also    ->    See_also
+```
+
+The exceptions are `prose.short_description`, `prose.description?`, and `prose.*`.
+
+#### prose.short_description
+
+This ingredient isn't demarcated by an `<h2>` at all. To satisfy the ingredient, a page must:
+
+- contain a `<p>` element
+- that is not a note or a warning
+- before the interactive example or the first `<h2>`, whichever comes first
+
+#### prose.description?
+
+This is an optional prose ingredient. To satisfy the ingredient, if the page contains an `h2#Description`, then it must appear in the place specified by the recipe.
```

Comment:
Should we explicitly say that the question mark makes ingredients optional? Maybe this duplicates what we have in the writing guide, so maybe we should just link to it somewhere, so that people can learn about recipes first, if needed? https://github.com/mdn/stumptown-content/blob/master/project-docs/stumptown-writers-guide.md#recipe-files

Reply:
I added a section on optional ingredients: https://github.com/mdn/stumptown-content/pull/391/commits/79b22c7931d8120fe1c3c1bc810517d6b623c4ed. Given all the ongoing weirdness around stumptown and Markdown, I want this doc to be a single self-contained item describing linting to recipes, so I don't really want to link out to general stumptown docs. (note that it does not mention Markdown or stumptown at all).

Reply:
Thanks Will! Agree it's a good idea to be self-contained. Also great you now talk about the fact that JS ingredients aren't optional in some JS recipes and mandatory in others. I thought this was an interesting discovery during our work in the last sprint.